python/ray/tune/checkpoint_manager.py

```diff
@@ -1,237 +1,113 @@
-
-
-class QueueItem:
-    def __init__(self, priority, value):
-        self.priority = priority
-        self.value = value
+        if keep_checkpoints_num == 0:
+            raise RuntimeError(
+                "If checkpointing is enabled, Ray Tune requires `keep_checkpoints_num` "
+                "to be None or a number greater than 0"
+            )
 
-    def __lt__(self, other):
-        return self.priority < other.priority
+        checkpoint_score_desc = checkpoint_score_attr.startswith("min-")
+        if checkpoint_score_desc:
+            checkpoint_score_attr = checkpoint_score_attr[4:]
+        else:
+            checkpoint_score_attr = checkpoint_score_attr
 
-    def __repr__(self):
-        return f"QueueItem({repr(self.value)})"
+        checkpoint_strategy = CheckpointStrategy(
+            num_to_keep=keep_checkpoints_num,
+            checkpoint_score_attribute=checkpoint_score_attr,
+            checkpoint_score_order=MIN if checkpoint_score_desc else MAX,
+        )
 
+        super().__init__(checkpoint_strategy=checkpoint_strategy, delete_fn=delete_fn)
 
-class CheckpointManager:
-    """Manages checkpoints on the driver for a trial."""
+    def on_checkpoint(self, checkpoint: _TrackedCheckpoint):
+        # Set checkpoint ID
+        checkpoint.checkpoint_id = (
+            checkpoint.checkpoint_id or self._latest_checkpoint_id
+        )
+        self._latest_checkpoint_id += 1
 
-    def __init__(
-        self,
-        keep_checkpoints_num: int,
-        checkpoint_score_attr: str,
-        delete_fn: Callable[[str], None],
-    ):
-        """Initializes a new CheckpointManager.
-
-        `newest_persistent_checkpoint` and `newest_memory_checkpoint` are
-        initialized to Checkpoint objects with values of None.
-
-        Args:
-            keep_checkpoints_num: Keep at least this many checkpoints.
-            checkpoint_score_attr: Attribute to use to determine which
-                checkpoints to keep.
-            delete_fn: Function that deletes checkpoints. Must be
-                idempotent.
-        """
-        self.keep_checkpoints_num = keep_checkpoints_num or float("inf")
-        assert (
-            self.keep_checkpoints_num > 0
-        ), "keep_checkpoints_num must be greater than 0."
-        self._checkpoint_score_desc = checkpoint_score_attr.startswith("min-")
-        if self._checkpoint_score_desc:
-            self._checkpoint_score_attr = checkpoint_score_attr[4:]
+        if checkpoint.storage_mode == _TrackedCheckpoint.MEMORY:
+            self._replace_latest_memory_checkpoint(checkpoint)
         else:
-            self._checkpoint_score_attr = checkpoint_score_attr
+            assert checkpoint.storage_mode == _TrackedCheckpoint.PERSISTENT
+            assert (
+                self._checkpoint_strategy.num_to_keep is None
+                or self._checkpoint_strategy.num_to_keep > 0
+            )
+            self._decide_what_to_do_with_checkpoint(checkpoint)
```

Comment:
Ah no, these are two different constructs, especially within Tune. In Train yeah, a checkpoint is first memory and then persistent, but in Tune it can be either and we usually don't want to mix them up - at least not in the current refactoring where we want to keep all of the current functionality and contracts as they are.